### PR TITLE
Add supplemental error for `@system` variable access

### PR DESCRIPTION
**Pull request by dkorpel**
Suggested by @maxhaton https://github.com/dlang/dmd/pull/15434#issuecomment-1652541285

### CODE REVIEW

File: compiler/src/dmd/expressionsem.d [hunk 1]
```diff
@@ -875,6 +875,13 @@ Lagain:
                 if (sc.setUnsafePreview(global.params.systemVariables, false, loc,
                     "cannot access `@system` variable `%s` in @safe code", sd))
                 {
+                    if (auto v = sd.isVarDeclaration())
+                    {
+                        if (v.systemInferred)
+                            errorSupplemental(v.loc, "`%s` is inferred to be `@system` from its initializer", v.toChars());
+                        else
+                            errorSupplemental(v.loc, "`%s` is located here", v.toChars());
```
Comment (SixthDot) [inline compiler/src/dmd/expressionsem.d:883]: how about _declared_ instead of located ?

Comment (dkorpel) [reply compiler/src/dmd/expressionsem.d]: Done

**Comment by maxhaton** [comment]
Thx. Will review later tonight

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
@maxhaton any objections?

File: compiler/src/dmd/expressionsem.d [hunk 2]
```diff
@@ -875,6 +875,13 @@ Lagain:
                 if (sc.setUnsafePreview(global.params.systemVariables, false, loc,
                     "cannot access `@system` variable `%s` in @safe code", sd))
                 {
+                    if (auto v = sd.isVarDeclaration())
+                    {
+                        if (v.systemInferred)
+                            errorSupplemental(v.loc, "`%s` is inferred to be `@system` from its initializer here", v.toChars());
```
Comment (SixthDot) [inline compiler/src/dmd/expressionsem.d:881]: The if-else statement can be replaced with a conditional expression in the call to errorSupplemental:

```
errorSupplemental(v.loc, v.systemInferred ? "`%s` is inferred to be `@system` from its initializer here" : "`%s` is declared here", v.toChars());
```

**Comment by maxhaton** [comment]
ok to me

**Comment by RazvanN7** [comment]
@dkorpel a rebase will fix the buildkite failure. Also, I leave it up to you if you consider @SixthDot's comment.

File: compiler/src/dmd/expressionsem.d (hunk 2 above)
Comment (dkorpel) [reply compiler/src/dmd/expressionsem.d:881]: I think it reads better with each error having its own statement

Comment (SixthDot) [reply compiler/src/dmd/expressionsem.d:881]: expressions are cheaper, here the statement introduces a new layer of semantic analysis that is of no utility.

**Pull request merged**